### PR TITLE
support delegate singleton object&add compress index&others useful function

### PR DESCRIPTION
**Pull request by owen200008**
1. assets-manager support compressindex
2.node object add noupdatecomponent,  
(when i write the richtext create 20000 node it use more time to do nothing just foreach the component, most component is no need update)
3.node add destroy event to lua(js i no use, so please check it is good)
(it is use to some node add or delete from father but not delete object, we should know the real destroy event to delete other resource)
4.listview add changeitemevent and support touch location
(it is useful and easy hightlight the item when choose, end event set unenable ,changeevent set enable,
location is useful to bussiness to show tip for the item)
5.ResizableBuffer need length it is useful
6.singleton object support setdelegate like file, it is useful to bussiness, ResizableBuffer length is use for file to decode filedata when the fieldata is encode
7.fast tmx support multi tileset and read resource with plist and support straggered

### CODE REVIEW

**Comment by DavidDeSimone** [comment]
As a stylist note, it looks like you are indenting with tabs. Please indent with spaces for project consistency.

**Comment by owen200008** [comment]
ok get it. i have change it and fix bug and export the interface for lua.

File: cocos/2d/CCSpriteFrameCache.cpp [hunk 1]
```diff
@@ -60,6 +60,14 @@ SpriteFrameCache* SpriteFrameCache::getInstance()
     return _sharedSpriteFrameCache;
 }
 
+void SpriteFrameCache::setDelegate(SpriteFrameCache* pCache)
+{
+    swap(_sharedSpriteFrameCache, pCache);
+    if (pCache){
```
Comment (j-jorge) [inline cocos/2d/CCSpriteFrameCache.cpp]: `CC_SAFE_RELEASE_NULL` tests the pointer for nullness so it's useless to test it here.

File: cocos/ui/UIListView.cpp [hunk 2]
```diff
@@ -540,12 +540,38 @@ void ListView::selectedItemEvent(TouchEventType event)
...
                 _ccEventCallback(this, static_cast<int>(EventType::ON_SELECTED_ITEM_END));
             }
+            //end calc
+            if (event == TouchEventType::ENDED)
+            {
+                //only ended will know last choose one，begin just highlight
+                Widget* parent = sender;
+                while (parent)
+                {
+                    if (parent && (parent->getParent() == _innerContainer))
```
Comment (j-jorge) [inline cocos/ui/UIListView.cpp]: `parent` is necessarily not null here. The test should be `if (parent->getParent() == _innerContainer)`.

File: cocos/2d/CCSpriteFrameCache.cpp (hunk 1 above)
Comment (owen200008) [reply cocos/2d/CCSpriteFrameCache.cpp]: ok i will change

File: cocos/ui/UIListView.cpp (hunk 2 above)
Comment (owen200008) [reply cocos/ui/UIListView.cpp]: ok i will change

**Comment by minggo** [comment]
@owen200008 thanks for your contribution, but it is a huge PR, and do many different things in one PR. Could you please break it into more small ones? Just do one thing in a small PR, and it is better to add a test case for new feature or bug fixing. Thanks.

**Comment by owen200008** [comment]
ok i will try. i'm not familiar with git push. so maybe it will be slow.

**Comment by minggo** [comment]
thanks

**Comment by minggo** [comment]
@owen200008 i close this PR since you will create more small PRS.

**Pull request closed without merging**